### PR TITLE
Introduce workspaces

### PR DESCRIPTION
**Pull request by BanzaiMan**
Workspaces allow jobs _*within*_ a build to share files from one stage to the next.

### CODE REVIEW

**Comment by BanzaiMan** [comment]
Depends on
- travis-ci/casher#46 
- travis-ci/travis-scheduler#189

**Comment by joshk** [comment]
@BanzaiMan are the PRs for casher and scheduler backwards compatible? eg. should we get those merged and deployed?

Also, do you have a build on staging to demo this on?

Annnnnd are we using the Travis repo ID or GitHub ID in the urls?

**Comment by BanzaiMan** [comment]
casher PR is backward compatible, so it hasn't been tested that way.

scheduler PR is; if the config setting doesn't exist, it'll just be ignored.

Neither Travis repo ID nor GitHub ID is used; just the build ID.

**Comment by BanzaiMan** [comment]
Needs `travis-yml` work to correspond to the change.

File: lib/travis/build/script/shared/directory_cache/base.rb [hunk 1]
```diff
@@ -192,14 +192,15 @@ def validate
             end
 
             def run(command, args, options = {})
+              cache_name = URI.encode_www_form_component(slug)
               sh.with_errexit_off do
                 sh.if "-f #{BIN_PATH}" do
                   sh.if "-e ~/.rvm/scripts/rvm" do
                     sh.cmd('type rvm &>/dev/null || source ~/.rvm/scripts/rvm', echo: false, assert: false)
-                    sh.cmd "rvm $(travis_internal_ruby) --fuzzy do #{BIN_PATH} #{command} #{Array(args).join(' ')}", options.merge(echo: false, assert: false)
+                    sh.cmd "rvm $(travis_internal_ruby) --fuzzy do #{BIN_PATH} --name #{cache_name} #{archive_type} #{command} #{Array(args).join(' ')}", options.merge(echo: false, assert: false)
```
Comment (BanzaiMan) [inline lib/travis/build/script/shared/directory_cache/base.rb]: Probably we should extract constructing the `casher` command to a method.

**Comment by joshk** [comment]
@BanzaiMan lets have a call/catchup on Monday about the next steps to getting this merged and more widely tested

**Comment by BanzaiMan** [comment]
travis-ci/casher#46 in production → https://travis-ci.org/BanzaiMan/travis_production_test/builds/549861457 (backwards compat)

**Pull request merged**